lisa/sut_orchestrator/azure/common.py

```diff
@@ -2065,17 +2089,121 @@ def check_blob_exist(
-) -> bool:
+    raise_error: bool = True,
+) -> None:
     container_client = get_or_create_storage_container(
         credential=platform.credential,
         subscription_id=platform.subscription_id,
         cloud=platform.cloud,
         account_name=account_name,
         container_name=container_name,
         resource_group_name=resource_group_name,
+        allow_create=False,
+    )
+    blob_client = container_client.get_blob_client(blob_name)
+    blob_exist = blob_client.exists()
+    if raise_error and not blob_exist:
+        raise LisaException(f"Blob {blob_name} does not exist.")
+
+
+def download_blob(
+    account_name: str,
+    container_name: str,
+    blob_name: str,
+    file_path: Path,
+    log: Logger,
+    cloud: Cloud = AZURE_PUBLIC_CLOUD,
+    credential: Optional[Any] = None,
+    subscription_id: Optional[str] = None,
+    resource_group_name: Optional[str] = None,
+    connection_string: Optional[str] = None,
+) -> PurePath:
+    container_client = get_or_create_storage_container(
+        container_name,
+        credential,
+        subscription_id,
+        cloud,
+        account_name,
+        resource_group_name,
+        connection_string,
+        False,
     )
     blob_client = container_client.get_blob_client(blob_name)
-    return blob_client.exists()
+    if not blob_client.exists():
+        raise LisaException(f"Blob {blob_name} not found in container {container_name}")
+    try:
+        with _global_download_blob_lock:
+            blob_properties = blob_client.get_blob_properties()
+            if file_path.exists():
+                blob_hash = blob_properties.content_settings.content_md5
+                if blob_hash:
+                    blob_hash_hex = "".join(f"{byte:02x}" for byte in blob_hash)
+                    if _calculate_hash(file_path) == blob_hash_hex:
+                        log.debug(
+                            f"Blob {blob_name} already exists in {file_path}. "
+                            "No need to download again."
+                        )
+                        return file_path
+
+            blob_size = blob_properties["size"]
+            downloaded_size = 0
+            start_time = time()
+            log_interval = 10
+            next_log_time = start_time + log_interval
+            with open(file_path, "wb") as file:
+                download_stream = blob_client.download_blob()
+                for chunk in download_stream.chunks():
+                    file.write(chunk)
+                    downloaded_size += len(chunk)
+                    percentage = (downloaded_size / blob_size) * 100
+                    current_time = time()
+                    if current_time >= next_log_time:
+                        log.debug(
+                            f"Downloaded {downloaded_size}/{blob_size} bytes "
+                            f"({percentage:.2f}% complete)"
+                        )
+                        next_log_time += log_interval
```

Comment:
The next logging time should be at least 10 seconds from now.
`next_log_time = current_time + log_interval`

Reply:
updated